core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -180,14 +198,93 @@ class TransactionCoordinator(brokerId: Int,
+      responseCallback(errors)
+    else
+      txnManager.getTransaction(transactionalId) match {
+        case None =>
+          responseCallback(Errors.INVALID_PID_MAPPING)
+        case Some(metadata) =>
+          metadata synchronized {
+            if (metadata.pid != pid)
+              responseCallback(Errors.INVALID_PID_MAPPING)
+            else if (metadata.epoch != epoch)
+              responseCallback(Errors.PRODUCER_FENCED)
+            else metadata.state match {
+              case Ongoing =>
+                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+              case CompleteCommit =>
+                if (command == TransactionResult.COMMIT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case CompleteAbort =>
+                if (command == TransactionResult.ABORT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case _ =>
+                responseCallback(Errors.INVALID_TXN_STATE)
+            }
+          }
+      }
+  }
+
+  private def commitOrAbort(transactionalId: String,
+                            pid: Long,
+                            epoch: Short,
+                            command: TransactionResult,
+                            responseCallback: EndTxnCallback,
+                            metadata: TransactionMetadata) = {
+    val nextState = if (command == TransactionResult.COMMIT) PrepareCommit else PrepareAbort
+    val newMetadata = new TransactionMetadata(pid, epoch, metadata.txnTimeoutMs, nextState, metadata.topicPartitions, time.milliseconds())
+    metadata.prepareTransitionTo(nextState)
+
+    def logAppendCallback(errors: Errors): Unit = {
+      if (errors != Errors.NONE)
+      // TODO: What do we do in this case?
+        responseCallback(errors)
+      else {
+        txnManager.coordinatorEpochFor(transactionalId) match {
+          case Some(coordinatorEpoch) =>
+            def completionCallback(error: Errors): Unit = {
+              val preparedCommitMetadata = txnManager.getTransaction(transactionalId).get
+              val completedState = if (nextState == PrepareCommit) CompleteCommit else CompleteAbort
+              val committedMetadata = new TransactionMetadata(pid,
+                epoch,
+                newMetadata.txnTimeoutMs,
+                completedState,
+                metadata.topicPartitions,
+                newMetadata.timestamp)
+              preparedCommitMetadata.prepareTransitionTo(completedState)
+              txnManager.appendTransactionToLog(transactionalId, committedMetadata, responseCallback, 1)
+            }
+            txnMarkerChannelManager.addTxnMarkerRequest(newMetadata, coordinatorEpoch, completionCallback)
+          case None =>
+            // TODO: What to do here? Is this possible at this point?
```

Comment:
It would mean that the tc isn't the coordinator for the partition anymore. Which would mean that the partition has emigrated during `handleEndTransaction`, which i guess is possible? note we validate earlier that this TC is the coordinator of the partition